File: p4/PartitionStream.cpp
// Jomikael Ruiz
// CPSC 3200
// PartitionStream.cpp
// Date Created: Oct. 6th, 2024
// Last Modified: Nov. 15th, 2024

#include "PartitionStream.h"

#include <memory>
#include "MsgStream.h"
using namespace std;

/**
 * pre-condition: partitionKey is a valid integer
 */
int PartitionStream::arrayIndexHash(int partitionKey)
{
	if(partitionKey < 0)
		throw(invalid_argument("partitionKey is negative"));
	int keyIndex = partitionKey % partitionSize;
	if(partitionKeyStream[keyIndex] == partitionKey)
		return keyIndex;
	for(int i = 0; i < partitionSize; i++) {
		if(partitionKeyStream[i] == partitionKey)
			return i;
		if(partitionKeyStream[i] == DEFAULT_KEY_VALUE) {
			return i;
		}
	}
	return DEFAULT_KEY_VALUE;
}

/**
 * pre-condition: partitionKey is a valid integer
 */
void PartitionStream::isValidKey(int partitionKey)
{
	int arrayIndex = arrayIndexHash(partitionKey);
	if( arrayIndex == DEFAULT_KEY_VALUE )
	{
		throw(invalid_argument("The partitionKey was not found in the list"));
	}
}

bool PartitionStream::isEmpty()
{
	return numberOfStreams == 0;
}

/**
 * pre-condition: messageStream is a valid MsgStream pointer,
 * 				  partitionKey is a valid integer
 * post-condition: messageStream is inserted into the internal array
 */
void PartitionStream::setMessageStream(int partitionKey, unique_ptr<MsgStream>& messageStream)
{
	if(!messageStream)
		throw(invalid_argument("MsgStream pointer is nullptr"));

	isValidKey(partitionKey);

	int arrayIndex = arrayIndexHash(partitionKey);

	this->partitionStream[arrayIndex] = *messageStream.release();
	this->partitionKeyStream[arrayIndex] = partitionKey;
	numberOfStreams++;
}

/**
 * pre-condition: src is a valid PartitionStream to deep copy from
 * post-condition: a deep copied version of src is created
 */
void PartitionStream::copyPartition(const PartitionStream& src)
{
	this->partitionSize = src.partitionSize;
	this->numberOfStreams = src.numberOfStreams;
	this->partitionStream = make_unique<MsgStream[]>(partitionSize);
	this->partitionKeyStream.reset(new int[partitionSize]);

	for( int i = 0; i < partitionSize; i++ )
	{
		if( partitionKeyStream[i] == DEFAULT_KEY_VALUE )
			continue;
		this->partitionStream[i] = src.partitionStream[i];
		this->partitionKeyStream[i] = src.partitionKeyStream[i];
	}
}


/**
 * post-condition: PartitionStream is completely reset to its original form
 */
void PartitionStream::deepReset()
{
	for( unsigned int i = 0; i < partitionSize; i++ )
	{
		try {
			partitionStream[i].resetStream();
		} catch(exception& out_of_range) {
			continue;
		}
	}
	partitionStream.reset();
	partitionKeyStream.reset();
}

/**
 * pre-condition: PartitionStream's paritionSize must not be at MAX_MSGSTREAM_OBJECTS
 * post-condition: partitionSize is incrememented by one and pointers are resized
 */
void PartitionStream::increaseSize()
{
	if(partitionSize >= MAX_MSGSTREAM_OBJECTS)
		throw(out_of_range("PartitionStream at Max Size"));
	this->partitionSize++;
	unique_ptr<MsgStream[]> tempStream = make_unique<MsgStream[]>(partitionSize);
	unique_ptr<int[]> tempKeyStream = make_unique<int[]>(partitionSize);
	for(int i = 0; i < partitionSize - 1; i++) {
		tempStream[i] = partitionStream[i];
		tempKeyStream[i] = partitionKeyStream[i];
	}
	tempKeyStream[partitionSize - 1] = DEFAULT_KEY_VALUE;
	partitionStream.swap(tempStream);
	partitionKeyStream.swap(tempKeyStream);
}

/**
 * pre-condition: PartitionStream must not have the last array index occupied
 * post-condition: partitionSize is decremented by one and pointers are resized
 */
void PartitionStream::decreaseSize()
{
	if(partitionKeyStream[partitionSize - 1] != DEFAULT_KEY_VALUE)
		throw (out_of_range("A partition is occupying the last index, cannot decrement size."));
	this->partitionSize--;
	unique_ptr<MsgStream[]> tempStream = make_unique<MsgStream[]>(partitionSize);
	unique_ptr<int[]> tempKeyStream = make_unique<int[]>(partitionSize);
	for(int i = 0; i < partitionSize; i++) {
		tempStream[i] = partitionStream[i];
		tempKeyStream[i] = partitionKeyStream[i];
	}
	partitionStream.swap(tempStream);
	partitionKeyStream.swap(tempKeyStream);
}

/**
 * pre-condition: stream must be a valid PartitionStream
 */
bool PartitionStream::checkEquality(const PartitionStream& stream) {
	if (stream.partitionSize != this->partitionSize)
		return false;
	for(int i = 0; i < partitionSize; i++) {
		if(stream.partitionKeyStream[i] == DEFAULT_KEY_VALUE
			&& this->partitionKeyStream[i] == DEFAULT_KEY_VALUE)
			continue;

		if(stream.partitionKeyStream[i] == DEFAULT_KEY_VALUE
		   || this->partitionKeyStream[i] == DEFAULT_KEY_VALUE)
			return false;

		bool equalityCheck = ((stream.partitionStream[i])==(this->partitionStream[i]));
		if(!equalityCheck) {
			return false;
		}
	}
	return true;
}

/**
 * pre-condition: size must be a valid positive integer
 * post-condition: PartitionStream's data members are initialized
 */
void PartitionStream::initializeStream(int size)
{
	if(size <= 0) {
		throw(invalid_argument("non-positive input size"));
	}
	partitionSize = size;
	partitionStream = make_unique<MsgStream[]>(partitionSize);
	partitionKeyStream = make_unique<int[]>(partitionSize);
	for (int i = 0; i < partitionSize; i++)
	{
		partitionKeyStream[i] = DEFAULT_KEY_VALUE;
	}
	numberOfStreams = INITIAL_VALUE;
}

PartitionStream::PartitionStream()
{
	initializeStream(DEFAULT_MSGSTREAM_OBJECTS);
}

PartitionStream::PartitionStream(int size)
{
	initializeStream(size);
}

/**
 * pre-condition: messageStream is a valid MsgStream * for dependency injection
 * 				  partitionKey is a valid integer
 */
PartitionStream::PartitionStream(int partitionKey, unique_ptr<MsgStream> messageStream)
{
	initializeStream(DEFAULT_MSGSTREAM_OBJECTS);
	setMessageStream(partitionKey, messageStream);
}

/**
 * pre-condition: src must be a valid PartitionStream
 */
PartitionStream::PartitionStream(const PartitionStream& src)
{
	copyPartition(src);
}

/**
 * pre-condition: src is a valid PartitionStream
 */
PartitionStream::PartitionStream(PartitionStream&& src) noexcept
{
	this->partitionSize = src.partitionSize;
	this->partitionStream.reset(src.partitionStream.release());
	this->partitionKeyStream.reset(src.partitionKeyStream.release());

	src.partitionSize = 0;
}

/**
 * pre-condition: src is a valid PartitionStream object
 * post-condition: PartitionStream object is deep copied from src
 */
PartitionStream& PartitionStream::operator=(const PartitionStream& src)
{
	if( this == &src )
		return *this;

	deepReset();
	copyPartition(src);

	return *this;
}

/**
 * pre-condition: src is a valid PartitionStream object
 * post-condition: PartitionStream object is deep copied from src
 */
PartitionStream& PartitionStream::operator=(PartitionStream&& src) noexcept
{
	if( this == &src )
		return *this;

	swap(this->partitionSize, src.partitionSize);
	swap(this->partitionStream, src.partitionStream);
	swap(this->partitionKeyStream, src.partitionKeyStream);

	return *this;
}

/**
 * pre-condition: partitionKey is a valid positive integer
 * post-condition: MsgStream at partitionKey is returned
 */
MsgStream& PartitionStream::operator[](int partitionKey)
{
	isValidKey(partitionKey);
	int index = arrayIndexHash(partitionKey);
	if(partitionKeyStream[index] == DEFAULT_KEY_VALUE) {
		partitionKeyStream[index] = partitionKey;
		numberOfStreams++;
	}
	return partitionStream[index];
}

/**
 * pre-condition: stream is a valid PartitionStream
 */
bool PartitionStream::operator==(const PartitionStream& stream)
{
	return checkEquality(stream);
}

/**
 * pre-condition: stream is a valid PartitionStream
 */
bool PartitionStream::operator!=(const PartitionStream& stream)
{
	return !checkEquality(stream);
}

/**
 * pre-condition: stream is a valid PartitionStream
 */
bool PartitionStream::operator>=(const PartitionStream& stream) const
{
	if(this->numberOfStreams >= stream.numberOfStreams)
		return true;
	return false;
}

/**
 * pre-condition: stream is a valid PartitionStream
 */
bool PartitionStream::operator>(const PartitionStream& stream) const
{
	if(this->numberOfStreams > stream.numberOfStreams)
		return true;
	return false;
}

/**
 * pre-condition: stream is a valid PartitionStream
 */
bool PartitionStream::operator<=(const PartitionStream& stream) const
{
	if(this->numberOfStreams <= stream.numberOfStreams)
		return true;
	return false;
}

/**
 * pre-condition: stream is a valid PartitionStream
 */
bool PartitionStream::operator<(const PartitionStream& stream) const
{
	if(this->numberOfStreams < stream.numberOfStreams)
		return true;
	return false;
}

/**
 * pre-condition: stream is a valid PartitionStream,
 * 				  the sum of both numberOfStreams between objects must not be > MAX_MSG_OBJECTS
 * post-condition: a temporary PartitionStream with all partitions between two objects is returned
 */
PartitionStream PartitionStream::operator+(PartitionStream& stream)
{
	if(stream.numberOfStreams + this->numberOfStreams > MAX_MSGSTREAM_OBJECTS)
		throw(out_of_range("Size of PartitionStreams exceeds maxSize"));

	PartitionStream temp(*this);
	unsigned int newSize = stream.numberOfStreams + temp.numberOfStreams;
	while(temp.partitionSize <= newSize)
		temp++;
	for(int i = 0; i < stream.partitionSize; i++) {
		if(stream.partitionKeyStream[i] == DEFAULT_KEY_VALUE)
			continue;
		MsgStream::ArrayReturn messages = stream[stream.partitionKeyStream[i]].readAllMessage();
		for(int j = 0; j < messages.streamSize; j++) {
			temp.appendMessage(stream.partitionKeyStream[i], messages.messages[j]);
		}
	}
	return temp;
}

/**
 * pre-condition: stream is a valid PartitionStream,
 * 				  the sum of both numberOfStreams between objects must not be > MAX_MSG_OBJECTS
 * post-condition: a temporary PartitionStream with all partitions between two objects is returned
 */
PartitionStream PartitionStream::operator+=(PartitionStream& stream)
{
	*this = *this + stream;
	return *this;
}

/**
 * pre-condition: PartitionStream's paritionSize must not be at MAX_MSGSTREAM_OBJECTS
 * post-condition: partitionSize is incrememented by one and pointers are resized
 */
PartitionStream PartitionStream::operator++()
{
	increaseSize();
	return *this;
}

/**
 * pre-condition: PartitionStream's paritionSize must not be at MAX_MSGSTREAM_OBJECTS
 * post-condition: partitionSize is incrememented by one and pointers are resized
 */
PartitionStream PartitionStream::operator++(int input)
{
	PartitionStream oldState = *this;
	increaseSize();
	return oldState;
}

/**
 * pre-condition: PartitionStream must not have the last array index occupied
 * post-condition: partitionSize is decremented by one and pointers are resized
 */
PartitionStream PartitionStream::operator--()
{
	decreaseSize();
	return *this;
}

/**
 * pre-condition: PartitionStream must not have the last array index occupied
 * post-condition: partitionSize is decremented by one and pointers are resized
 */
PartitionStream PartitionStream::operator--(int input)
{
	PartitionStream oldState = *this;
	decreaseSize();
	return oldState;
}

/**
 * post-condition: PartitionStream is completely deallocated
 */
PartitionStream::~PartitionStream()
{
	deepReset();
}

/**
 * pre-condition: partitionKey is a valid integer, message is a valid string
 * post-condition: string is added to the respective partition
 */
void PartitionStream::appendMessage(int partitionKey, string message)
{
	isValidKey(partitionKey);
	int arrayIndex = arrayIndexHash(partitionKey);
//	if( !this->partitionStream[arrayIndex] )
//	{
//		partitionStream[arrayIndex] = make_unique<MsgStream>();
//		partitionKeyStream[arrayIndex] = partitionKey;
//	}
	if(this->partitionKeyStream[arrayIndex] == DEFAULT_KEY_VALUE) {
		partitionKeyStream[arrayIndex] = partitionKey;
		numberOfStreams++;
	}

	partitionStream[arrayIndex].appendMessage(message);
}

/**
 * pre-condition: partitionKey is a valid integer, messages is a valid MsgStream ptr
 * post-condition: string is added to the respective partition
 */
void PartitionStream::appendMessage(int partitionKey, unique_ptr<MsgStream>& messages)
{
	if( !messages )
		throw( invalid_argument("MsgStream pointer is nullptr") );
	setMessageStream(partitionKey, messages);
}

/**
 * pre-condition: PartitionStream must not be empty to reset,
 * 				  otherwise an exception is thrown
 * post-condition: internal array is completely wiped and reset
 */
void PartitionStream::resetStream()
{
	if( isEmpty() )
		throw(invalid_argument("The Stream is already empty"));

	deepReset();
	partitionStream = make_unique<MsgStream[]>(partitionSize);
	partitionKeyStream = make_unique<int[]>(partitionSize);
	for(int i = 0; i < partitionSize; i++)
		partitionKeyStream[i] = DEFAULT_KEY_VALUE;
	numberOfStreams = INITIAL_VALUE;
}

/**
 * pre-condition: the PartitionStream must not be empty,
 * 				  otherwise an exception is thrown,
 * 				  partitionKey must be a valid integer
 * post-condition: a duplicate array with all values in the respective partition
 * 				   is returned in the form of a structure,
 * 				   client is responsible for deallocation of string returned
 */
MsgStream::ArrayReturn PartitionStream::readMessages(int partitionKey)
{
	isValidKey(partitionKey);
	int arrayIndex = arrayIndexHash(partitionKey);
	return partitionStream[arrayIndex].readAllMessage();
}

File: P5/PartitionStream.cs
// Jomikael Ruiz
// CPSC 3200
// PartitionStream.cs
// Date Created: 10/21/24
// Last Modified: 12/2/24

namespace P5;

/**
 * Class Invariants:
 * - DefaultSize will always be 5 MessageStreams
 * - Max PartitionStream size is 11 MessageStreams
 * - PartitionStream size must be in range 0-11
 * - Client may place DurableStreams in a PartitionStream
 * - Partition Keys will be positive integers with a set of operations performed
 *   to map a partitionKey to its respective index
 * - Partition Keys cannot be negative, otherwise an exception is thrown
 * - PartitionStream can be inherited by other classes to use and expand upon its implementation
 * - PartitionStream inherits IDisposable and directly implements Dispose() to properly
 *   dispose of the internal contents of PartitionStream
 * Immutables:
 * - Once set, PartitionStream size cannot be altered
 */

public class PartitionStream : IDisposable
{
    private const int DefaultKeyValue = -1;
    
    protected const int MaxMsgstreamObjects = 11;
    protected const int DefaultMsgstreamObjects = 5;

    private MsgStream[] _partitionStream;
    private int[] _partitionKeyStream;

    private int _partitionSize;
    private bool _isEmpty;
    private bool _isDisposed;

    /**
     * PreCondition: partitionKey is a valid positive integer or 0
     */
    private int ArrayIndexHash(int partitionKey)
    {
        return partitionKey % _partitionSize;
    }

    /**
     * PreCondition: partitionKey is a valid positive integer or 0
     */
    private void IsValidKey(int partitionKey)
    {
        if (partitionKey < 0)
        {
            throw new IndexOutOfRangeException("PartitionKey is a negative value");
        }
        
        int arrayIndex = ArrayIndexHash(partitionKey);
        
        if( _partitionKeyStream[arrayIndex] != -1 &&
            _partitionKeyStream[arrayIndex] != partitionKey )
        {
            throw new IndexOutOfRangeException("An index is already occupying the expected partition");
        }
    }

    /**
     * PreCondition: messageStream is a valid MsgStream or DurableStream,
     * 				  partitionKey is a valid integer
     * PostCondition: messageStream is inserted into the internal array
     */
    private void SetMessageStream(int partitionKey, MsgStream messageStream)
    {
        IsValidKey(partitionKey);

        int arrayIndex = ArrayIndexHash(partitionKey);

        _partitionStream[arrayIndex] = messageStream;
        _partitionKeyStream[arrayIndex] = partitionKey; 
        _isEmpty = false;
    }

    /**
     * PreCondition: PartitionStream must not be empty, otherwise an exception
     *               is thrown
     * PostCondition: PartitionStream is completely reset to its original state by calling
     *                reset() for internal MsgStreams & DurableStreams
     */
    private void Reset()
    {
        if (_isEmpty)
        {
            throw new IndexOutOfRangeException("The stream is empty.");
        }
        for (int i = 0; i < _partitionSize; i++)
        {
            if (_partitionKeyStream[i] == DefaultKeyValue)
                continue;
            _partitionStream[i].Reset();
            _partitionKeyStream[i] = DefaultKeyValue;
        }
        _isEmpty = true;
        _isDisposed = false;
    }

    /**
     * PreCondition: initialSize is a valid positive integer
     * PostCondition: all internal data for PartitionStream is initialized
     */
    private void InitializePartitionStream(int initialSize)
    {
        _partitionSize = initialSize;
        _partitionStream = new MsgStream[_partitionSize];
        _partitionKeyStream = new int[_partitionSize];
        for (int i = 0; i < _partitionSize; i++)
        {
            _partitionKeyStream[i] = DefaultKeyValue;
        }
        _isEmpty = true;
    }
    
    /**
     * PreCondition: src must be a valid PartitionStream
     * PostCondition: all data from src is DeepCopied into this PartitionStream
     */
    public void DeepCopy(PartitionStream src) {
        _partitionSize = src._partitionSize;
        _isEmpty = src._isEmpty;
        _partitionStream = new MsgStream[_partitionSize];
        _partitionKeyStream = new int[_partitionSize];

        for( int i = 0; i < _partitionSize; i++ )
        {
            if (src._partitionKeyStream[i] == -1)
                continue;
            _partitionStream[i] = new MsgStream( src._partitionStream[i] );
            _partitionKeyStream[i] = src._partitionKeyStream[i];
        }
    }

    public PartitionStream()
    {
        InitializePartitionStream(DefaultMsgstreamObjects);
    }
    
    /**
     * PreCondition: src must be a valid PartitionStream
     */
    public PartitionStream(PartitionStream src)
    {
        DeepCopy(src);
    }
    
    /**
     * PreCondition: inputSize must be a valid positive integer
     */
    public PartitionStream(int inputSize)
    {
        if (inputSize < 0)
            throw new IndexOutOfRangeException("The input size is a negative value");

        if (inputSize > MaxMsgstreamObjects)
            throw new IndexOutOfRangeException("The input size is too large.");
        
        InitializePartitionStream(inputSize);
    }

    /**
     * PreCondition: partitionKey is a valid positive integer, otherwise
     *               an exception is thrown. messageStream must be a valid
     *               MessageStream or DurableStream
     */
    public PartitionStream(int partitionKey, MsgStream messageStream)
    {
        InitializePartitionStream(DefaultMsgstreamObjects);
        SetMessageStream(partitionKey, messageStream);
    }

    /**
     * PreCondition: message is a valid string, partitionKey is a valid integer,
     *               otherwise an exception is thrown
     * PostCondition: message is inserted into the internal array
     */
    public virtual void AppendMessage(int partitionKey, string message)
    {
        if(_isDisposed)
            throw new ObjectDisposedException("The stream has been disposed.");
        int arrayIndex = ArrayIndexHash(partitionKey);

        IsValidKey(partitionKey);

        if (_partitionKeyStream[arrayIndex] == DefaultKeyValue)
        {
            _partitionStream[arrayIndex] = new MsgStream();
            _partitionKeyStream[arrayIndex] = partitionKey;
        }
        _partitionStream[arrayIndex].AppendMessage(message);
        _isEmpty = false;
    }

    /**
     * PreCondition: messageStream is a valid MsgStream or DurableStream,
     * 				  partitionKey is a valid integer
     * PostCondition: messageStream is inserted into the internal array
     */
    public virtual void AppendMessage(int partitionKey, MsgStream messageStream)
    {
        if(_isDisposed)
            throw new ObjectDisposedException("The stream has been disposed.");
        SetMessageStream(partitionKey, messageStream);
    }

    /**
     * PreCondition: PartitionStream must not be empty, otherwise an exception
     *               is thrown
     * PostCondition: PartitionStream is completely reset to its original state
     */
    public virtual void ResetStream()
    {
        Reset();
    }

    /**
     * PreCondition: partitionKey is a valid positive integer,
     *               otherwise an exception is thrown. PartitionStream
     *               must not be empty, otherwise an exception is thrown
     * PostCondition: a string array with the contents of the specific partition
     *                are returned
     */
    public string[] ReadMessages(int partitionKey)
    {
        if (_isEmpty)
            throw new IndexOutOfRangeException("The stream is empty.");
        IsValidKey(partitionKey);
        int arrayIndex = ArrayIndexHash(partitionKey);

        return _partitionStream[arrayIndex].ReadAllMessages();
    }

    /**
     * PreCondition: _partitionStream must have some MsgStreams/DurableStreams in
     *               its contents
     * PostCondition: each MsgStream/DurableStream within PartitionStream has their
     *                respective Dispose method called
     */
    public void Dispose()
    {
        Dispose(disposing: true);
        GC.SuppressFinalize(this);
    }

    /**
     * PreCondition: _partitionStream must have some MsgStreams/DurableStreams in
     *               its contents, _isDisposed must not be true
     * PostCondition: each MsgStream/DurableStream within PartitionStream has their
     *                respective Dispose method called
     */
    public virtual void Dispose(bool disposing)
    {
        if (_isDisposed)
            throw new ObjectDisposedException("The stream has been disposed.");
        if (disposing)
        {
            for (int i = 0; i < _partitionSize; i++)
            {
                _partitionStream[i].Dispose();
            }
        }
        _isDisposed = true;
    }

    /**
     * PostCondition: PartitionStream is cleaned up by the garbage collection
     */
    ~PartitionStream()
    {
        Dispose(disposing: false);
    }
};

/*
 * Implementation Invariants:
 * - When collisions occur internally with the
 *   partitionKey and arrayIndex, an exception is thrown
 * - Reading messages based on a partitionKey returns all
 *   the messages associated with that specific partition
 * - Object supports Deep Copy via deepCopy() method
 * - ShallowCopy is not supported within this class
 * - To return messages, a structure is returned containing the pointer to
 *   the array of messages and the size for easy traversal
 * - The Key Array is instantly set to values of -1.
 * - PartitionStream supports storage of both MsgStream and DurableStream
 * - Dispose() will simply call the Dispose() method for MsgStream and DurableStream
 *   to properly dispose of each stream within its member data array
 * - Dispose() cannot be undone unless, ResetStream() is called, resetting the state of PartitionStream
 */
